sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/RecordWriterManager.java

```diff
@@ -0,0 +1,289 @@
+ * Assigns one {@link DestinationState} per windowed destination. A {@link DestinationState} assigns
+ * one writer per partition in table destination. If the Iceberg {@link Table} is un-partitioned,
+ * the data is written normally using one {@link RecordWriter} (i.e. the {@link DestinationState}
+ * has one writer). At any given moment, the number of open data writers should be less than or
+ * equal to the number of total partitions (across all windowed destinations).
+ *
+ * <p>A {@link DestinationState} maintains its writers in a {@link Cache}. If a {@link RecordWriter}
+ * is inactive for 1 minute, the {@link DestinationState} will automatically close it to free up
+ * resources. Calling {@link #close()} on this {@link RecordWriterManager} will do the following for
+ * each {@link DestinationState}:
+ *
+ * <ol>
+ *   <li>Close all underlying {@link RecordWriter}s
+ *   <li>Collect all {@link DataFile}s
+ *   <li>Create a new {@link ManifestFile} referencing these {@link DataFile}s
+ * </ol>
+ *
+ * <p>After closing, the resulting {@link ManifestFile}s can be retrieved using {@link
+ * #getManifestFiles()}.
+ */
+class RecordWriterManager {
+  private static final Logger LOG = LoggerFactory.getLogger(RecordWriterManager.class);
+
+  /**
+   * Represents the state of one Iceberg table destination. Creates one {@link RecordWriter} per
+   * partition and manages them in a {@link Cache}.
+   *
+   * <p>On closing, each writer's output {@link DataFile} is collected.
+   */
+  class DestinationState {
+    private final IcebergDestination icebergDestination;
+    private final PartitionSpec spec;
+    private final org.apache.iceberg.Schema schema;
+    private final PartitionKey partitionKey;
+    private final String tableLocation;
+    private final FileIO fileIO;
+    private final String stateToken = UUID.randomUUID().toString();
+    private final List<DataFile> dataFiles = Lists.newArrayList();
+    @VisibleForTesting final Cache<PartitionKey, RecordWriter> writers;
+    @VisibleForTesting final Map<PartitionKey, Integer> writerCounts = Maps.newHashMap();
+
+    DestinationState(IcebergDestination icebergDestination, Table table) {
+      this.icebergDestination = icebergDestination;
+      this.schema = table.schema();
+      this.spec = table.spec();
+      this.partitionKey = new PartitionKey(spec, schema);
+      this.tableLocation = table.location();
+      this.fileIO = table.io();
+
+      // build a cache of RecordWriters.
+      // writers will expire after 1 min of idle time.
+      // when a writer expires, its data file is collected.
+      this.writers =
+          CacheBuilder.newBuilder()
+              .expireAfterAccess(1, TimeUnit.MINUTES)
+              .removalListener(
+                  (RemovalNotification<PartitionKey, RecordWriter> removal) -> {
+                    final RecordWriter recordWriter =
+                        Preconditions.checkNotNull(removal.getValue());
```

Comment:
Recommend using Beam's fork of `Preconditions` which raises correct exceptions and differentiates `IllegalStateException` and `IllegalArgumentException`. Guava is incorrect - it just raises NPE which should only be raised when the code in question has failed, not in response to bad data.

Reply:
Thanks, I substituted it for our `checkStateNotNull()`

I still needed Guava's Preconditions because Beam doesn't have equivalents for  `checkArgument()`, `checkState()`

Reply:
👍  those Guava methods throw the right kind of exception so they are g2g